### PR TITLE
Add vardiff test framework

### PR DESCRIPTION
**Pull request by Shourya742**
Partially solves: #1725

### CODE REVIEW

File: protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        );
+        assert!(
+            diff_percentage < 20.0,
+            "Change should be minimal for stable rate if any"
+        );
+        assert_eq!(vardiff.shares_since_last_update(), 0)
+    } else {
+        assert!(
+            true,
+            "No update for stable hashrate is acceptable if deviation is low"
```
Comment (plebhash) [inline protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs]: I don't understand this... why `assert!` on `true`?

the sentence in the string is also not clear, are we saying that if deviation is low, updates are unacceptable? or that it's acceptable that no updates happened?

and overall, do we expect an update to happen or not?

Comment (plebhash) [reply protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs]: and since hashrate didn't change, shouldn't we simply assert that `vardiff.try_vardiff()` returns `Ok(None)`?

I mean, I understand the API was designed to be robust against statistically insignificant changes, but the execution of this test is deterministic so I don't get why we would leave different `if / else` statements here?

File: protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        "Should not update within 15s; elapsed 10s"
+    );
+}
+
+pub fn test_try_vardiff_stable_hashrate_minimal_change_or_no_change<V: Vardiff>(vardiff: &mut V) {
+    let initial_hashrate = vardiff.hashrate();
+
+    let simulation_duration_secs = 70;
+    let expected_shares_for_duration =
+        (TEST_SHARES_PER_MINUTE * simulation_duration_secs as f32 / 60.0).round() as u32;
```
Comment (plebhash) [inline protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs]: given the linearity of this operation, what's the difference between setting a 70s duration and something much smaller, like 5s?

File: protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+use std::{thread, time::Duration};
+
+use crate::VardiffState;
+mod classic;
+
+use super::Vardiff;
+
+pub const TEST_INITIAL_HASHRATE: f32 = 1000.0;
+pub const TEST_SHARES_PER_MINUTE: f32 = 10.0;
+pub const TEST_MIN_ALLOWED_HASHRATE: f32 = 10.0;
```
Comment (plebhash) [inline protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs:11]: I did `cargo t` on this branch and I feel the test duration is a bit prohibitive.

waiting 70s or even 250s for a test to finish its execution feels a bit too much.

but also, I don't see any fundamental reason that forces us to do this

I mean, sure, expected share rate is set on a per-minute basis for UX reasons, but that doesn't mean that we must test scenarios for multiple minutes to assert the sanity of the APIs

if we remember that all math done here is linear, we can simply tweak these constants in order to allow us to perform our simulations under smaller execution timespans

Comment (plebhash) [reply protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs:11]: for example, if `TEST_SHARES_PER_MINUTE` is 60_000, we expect 1_000 shares per second

if in 1 second we only get 10 shares, that should be a trigger to lower the difficulty

if in 1 second we get 5_000 shares, that should be a trigger to raise the difficulty

if we want to emulate scenarios of "long" timespans, we can do ~10s

overall, we have plenty of room to play around with magnitudes, while trying to optimize for test execution time

File: protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        new_hashrate, TEST_MIN_ALLOWED_HASHRATE,
+        "Hashrate should be clamped to minimum"
+    );
+    assert_eq!(
+        vardiff.hashrate(),
+        TEST_MIN_ALLOWED_HASHRATE,
+        "Stored hashrate should be clamped"
+    );
+    assert_eq!(vardiff.shares_since_last_update(), 0);
+}
```
Comment (plebhash) [inline protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs]: as discussed in the call, it seems this test belongs in a different file?

File: protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        new_hashrate > initial_hashrate,
+        "Estimated hashratem should increase"
+    );
+    assert!(
+        vardiff.target() < initial_target,
+        "Target should become harder (smaller value)"
+    );
+    assert_eq!(vardiff.shares_since_last_update(), 0);
+}
+
```
Comment (plebhash) [inline protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs:183]: I see tests to assert target increase and decrease

but I don't see anything asserting that the vardiff converges to a correct hashrate estimation?

File: protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs (hunk 2 above)
Comment (plebhash) [reply protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs]: or even 1s?

File: protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs (hunk 3 above)
Comment (plebhash) [reply protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs:11]: ok I see the minimum window we have is 15s

https://github.com/stratum-mining/stratum/blob/e5fd8b7a35ba8f7c13e25608a8ec7f56116cb1d3/protocols/v2/roles-logic-sv2/src/vardiff/classic.rs#L157-L159

@GitGab19 I assume this is important for the correct behavior of the algorithm, right?

if so, then we can still make test durations much smaller, sticking to 15s as the default execution time and making artithmetic adaptations around `TEST_SHARES_PER_MINUTE`

Comment (bit-aloo) [reply protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs:11]: I’ve added a test configuration for this, which will also allow delta_time <= 0 in testing scenarios.

Comment (bit-aloo) [reply protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs:11]: Although the linearity seems tempting to exploit for testing, there’s a fundamental reason we need to wait for this duration our logic is designed to trigger at specific time intervals. Anyway, I’ve updated the tests to specifically target those wait points, and they’re now more intuitive to understand. The downside is that we still need to wait at least 60 seconds for those points to be reached. I initially considered adding some test-specific logic to bypass this, but that felt like heavy lifting given the current setup.

File: protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs (hunk 5 above)
Comment (bit-aloo) [reply protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs:183]: I understand your point about the convergence test. I’m planning to add that in a separate PR since it will require more work. This PR focuses on adding the basic skeleton for the testing framework for the vardiff module and strictly handling edge cases in the vardiff calculation. We already have a convergence mechanism in the difficulty management of tproxy, but it's somewhat unreliable. Also, simulating the entire hashing process is influenced by variable processing speeds, which makes it tricky. Designing such tests to be deterministic across the system would need a more careful and thorough review.

Comment (plebhash) [reply protocols/v2/roles-logic-sv2/src/vardiff/test/mod.rs:183]: when I wrote this comment originally, I wasn't necessarily thinking about dynamically testing with real hashrate from a CPU miner

my mindset was more around statically running the tests with specific values, such that we can assert that the vardiff algo converges into the right direction on its hashrate estimation

**Review by plebhash — APPROVED** [review]
